Source/Classes/Categories/NSURLSessionTask+Timing.m

```diff
@@ -11,111 +11,127 @@
 #import <objc/runtime.h>
 #import <QuartzCore/QuartzCore.h>
 
-static NSString * const kPINURLSessionTaskStateKey = @"state";
+static inline void pin_swizzleSelector(Class theClass, SEL originalSelector, SEL swizzledSelector) {
+    Method originalMethod = class_getInstanceMethod(theClass, originalSelector);
+    Method swizzledMethod = class_getInstanceMethod(theClass, swizzledSelector);
+    method_exchangeImplementations(originalMethod, swizzledMethod);
+}
 
-@interface PINURLSessionTaskObserver : NSObject
+static inline BOOL pin_addMethod(Class theClass, SEL selector, Method method) {
+    return class_addMethod(theClass, selector,  method_getImplementation(method),  method_getTypeEncoding(method));
+}
 
-@property (atomic, assign) CFTimeInterval startTime;
-@property (atomic, assign) CFTimeInterval endTime;
 
-- (instancetype)init NS_UNAVAILABLE;
-- (instancetype)initWithTask:(NSURLSessionTask *)task NS_DESIGNATED_INITIALIZER;
+@interface _PINURLSessionTaskSwizzling : NSObject
 
 @end
 
-@implementation PINURLSessionTaskObserver
+@implementation _PINURLSessionTaskSwizzling
 
-- (instancetype)initWithTask:(NSURLSessionTask *)task
-{
-    if (self = [super init]) {
-        _startTime = 0;
-        _endTime = 0;
-        [task addObserver:self forKeyPath:kPINURLSessionTaskStateKey options:0 context:nil];
++ (void)load {
+    /**
+     Swizzle NSURLSessionTask in AFNetworking's way
+     See AFURLSessionManager.m in AFNetworking project for detail
+     */
+    
+    if (NSClassFromString(@"NSURLSessionTask")) {
+        NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration ephemeralSessionConfiguration];
+        NSURLSession * session = [NSURLSession sessionWithConfiguration:configuration];
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wnonnull"
+        NSURLSessionDataTask *localDataTask = [session dataTaskWithURL:nil];
+#pragma clang diagnostic pop
+        IMP originalPINResumeIMP = method_getImplementation(class_getInstanceMethod([self class], @selector(PIN_resume)));
+        Class currentClass = [localDataTask class];
+        
+        while (class_getInstanceMethod(currentClass, @selector(resume))) {
+            Class superClass = [currentClass superclass];
+            
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wundeclared-selector"
+            IMP classResumeIMP = method_getImplementation(class_getInstanceMethod(currentClass, @selector(resume)));
+            Method afResumeMethod = class_getInstanceMethod(currentClass, @selector(af_resume));
```

Comment:
Looks like this code was copied from AFNetworking, I don't think their license allows that without also duplicating their license.

Reply:
Yeah, it is. I could change to a new style if method swizzling is still needed.